### PR TITLE
perf(editor): Add lint rules for optimization-friendly syntax

### PR DESCRIPTION
**Pull request by ivov**
https://linear.app/n8n/issue/CAT-321

### CODE REVIEW

File: packages/editor-ui/src/components/CodeNodeEditor/linter.ts [hunk 1]
```diff
@@ -559,6 +559,75 @@ export const useLinter = (
...
+				message: i18n.baseText('codeNodeEditor.linter.bothModes.dollarSignVariable'),
+			});
+		});
+
+		/**
+		 * Lint for $("myNode").item access in runOnceForAllItems mode
+		 *
+		 * $("myNode").item -> $("myNode").first()
+		 */
+		if (toValue(mode) === 'runOnceForEachItem') {
```
Comment (tomi) [inline packages/editor-ui/src/components/CodeNodeEditor/linter.ts:595]: Looks like we have some other rule that already covers this:

image

Maybe we should update the message in that case to mention that `.first()` could be used?

File: packages/editor-ui/src/components/CodeNodeEditor/linter.ts [hunk 2]
```diff
@@ -559,6 +559,75 @@ export const useLinter = (
 			});
 		});
 
+		/**
+		 * Lint for `$(variable)` usage where variable is not a string, in both modes.
+		 *
+		 * $(nodeName) -> <no autofix>
+		 */
+		const isDollarSignWithVariable = (node: Node) =>
```
Comment (tomi) [inline packages/editor-ui/src/components/CodeNodeEditor/linter.ts:567]: Looks like using a template literal needs to be handled separately (we have the same issue in the task runner as well):

image

File: packages/editor-ui/src/components/CodeNodeEditor/linter.ts [hunk 3]
```diff
@@ -559,6 +559,75 @@ export const useLinter = (
 			});
 		});
 
+		/**
```
Comment (tomi) [inline packages/editor-ui/src/components/CodeNodeEditor/linter.ts:562]: Do we have tests for this file?

**Review by tomi — COMMENTED** [review]
Couple comments 💬

File: packages/editor-ui/src/components/CodeNodeEditor/linter.ts (hunk 1 above)
Comment (ivov) [reply packages/editor-ui/src/components/CodeNodeEditor/linter.ts:596]: Good idea, added to this!

File: packages/editor-ui/src/components/CodeNodeEditor/linter.ts (hunk 2 above)
Comment (ivov) [reply packages/editor-ui/src/components/CodeNodeEditor/linter.ts:567]: Good catch, added checks for template string literals with and without vars.

File: packages/editor-ui/src/components/CodeNodeEditor/linter.ts (hunk 3 above)
Comment (ivov) [reply packages/editor-ui/src/components/CodeNodeEditor/linter.ts:562]: There's no tests for this file. Seeing as there's no setup already I decided to skip so I can move to other higher prio stories, else let me know if worth it and I'll see how to add them.

**Review by tomi — APPROVED** [review]
LGTM now 🚀

**Comment by janober** [comment]
Got released with [`n8n@1.68.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.68.0)

**Pull request merged**